### PR TITLE
Add fallback for external Qt in CMake file on Windows

### PR DESCRIPTION
**Pull request by jahorta**
On a clean build using CMake on Windows without Qt6 installed, CMakeLists does not use the external library for Qt 6.5.3 provided with the source. On windows, this PR adds a quiet check for Qt6, then if it is not available it adds the external library to the prefix path list and checks again for Qt6.

### CODE REVIEW

File: Source/CMakeLists.txt
Comment (cristian64) [inline Source/CMakeLists.txt:55]: Would this path be relative to the CMake file's location, or relative to the current working directory? We should aim for the former.

It may be a matter of prepending the path with `${CMAKE_CURRENT_LIST_DIR}\\`.

**Comment by cristian64** [comment]
The title of the PR may be a bit too vague. Suggestion:

> Add fallback for external Qt in CMake file on Windows.

**Comment by dreamsyntax** [comment]
Hi, rather than have merging commit please rebase this into one commit.
We are not as strict as dolphin but please follow the general commit guidelines.

You can do
`git rebase <branch>`
Or at this point it may be easier to `git cherry-pick` with your last commit on master.

File: Source/CMakeLists.txt
Comment (dreamsyntax) [inline Source/CMakeLists.txt:55]: I see CI built successfully, but isn't CMAKE_CURRENT_LIST_DIR going to resolve to the level where "\\..\\" isn't necessary?

**Review by dreamsyntax — CHANGES_REQUESTED** [review]

File: Source/CMakeLists.txt
Comment (cristian64) [reply Source/CMakeLists.txt:55]: The CI sets the `CMAKE_PREFIX_PATH` in the command line, so this code path is not being exercised in the CI right now.

Comment (cristian64) [reply Source/CMakeLists.txt:55]: > I see CI built successfully, but isn't CMAKE_CURRENT_LIST_DIR going to resolve to the level where "\..\" isn't necessary?

I would have thought it'd resolve to the `Source/` directory, therefore `\..\` would be necessary.

File: Source/CMakeLists.txt
Comment (cristian64) [inline Source/CMakeLists.txt:52]: Nit: I think a 4-space indentation is generally used in CMake files (and also in this file, except for a few poorly indented lines).

**Comment by cristian64** [comment]
> Hi, rather than have merging commit please rebase this into one commit. We are not as strict as dolphin but please follow the general commit guidelines.
> 
> You can do `git rebase <branch>` Or at this point it may be easier to `git cherry-pick` with your last commit on master.

I guess we can ultimately "squash and merge" to keep it in a single commit. Unless the author wants to tidy things up, which is generally my preference.

**Comment by jahorta** [comment]
I'm sorry. I don't really understand how git works very well and I think I just making it worse somehow... I tried to rebase, but it just added three more commits...

File: Source/CMakeLists.txt
Comment (dreamsyntax) [reply Source/CMakeLists.txt:55]: > > I see CI built successfully, but isn't CMAKE_CURRENT_LIST_DIR going to resolve to the level where "\..\" isn't necessary?
> 
> I would have thought it'd resolve to the `Source/` directory, therefore `\..\` would be necessary.

This may be correct, I didn't test it myself. I just wanted to ensure it is verified before merge.

**Comment by dreamsyntax** [comment]
> I'm sorry. I don't really understand how git works very well and I think I am just making it worse somehow... I tried to rebase, but it just added three more commits...

Rebasing would involve a force push. Guessing from the above you may have done it correctly but then tried to push, got the rejection advising to pull.

Instead do a rebase then force push
(`git push -f`)

**Comment by jahorta** [comment]
Okay, I was able to do the force-push, but now I don't know there is an extra commit (Update status icon with Dolphin's logo redesign). Not sure why it is showing up here since I forked the repository after this was committed...

**Comment by dreamsyntax** [comment]
You will need to rebase on the latest master to get around this.
You can:
`git checkout master`
`git pull`
`git checkout -` (the - brings you to prior branch)
`git rebase master`
`git push -f`

**Comment by dreamsyntax** [comment]
You may want to just start off with a fresh branch at this point and cherry pick your commits:
`git cherry-pick 6b1bab6097111b918200afaacfc6dbd5469a3f55` 
etc

**Comment by cristian64** [comment]
Note that you do not need to push to the remote until you have verified locally that the branch is based on the `master` branch and that it has a single commit with the expected changes.

**Comment by jahorta** [comment]
Maybe I misunderstood. Should I have all my changes in a single commit?

**Comment by dreamsyntax** [comment]
> Maybe I misunderstood. Should I have all my changes in a single commit?

In this case, yes. Since the changes are just updating to meet the repo standards. When working on a big feature it may make sense to have multiple commits. IMO for this we should just have one, with all your updates.

Your current push looks good to me, usually 1 commit is better but in this case its fine to use GH squash feature imo.

If you want to make your three commits one, you can squash them locally like this:

You can do:
`git rebase -i HEAD~3`

Changing the two latter commits from
`pick` to `squash` (or just letter s works)

Using this command will put you into vi/vim by default.
If you are not familiar here's a video with explanations and timestamps:
https://youtu.be/ggSyF1SVFr4

---
That said, what you have right now is fine. We can just use GitHub's squash button feature when merging. Purely just posting the above for learning purposes.

**Comment by cristian64** [comment]
> Maybe I misunderstood. Should I have all my changes in a single commit?

I think it's fine to have more than one commit in a branch. Commits need to be well polished with a properly written title and summary.

In this PR, I don't think we'd expect any more than one single commit with only changes in the CMake file.

**Comment by dreamsyntax** [comment]
Thanks for working on this, I hope we didn't overwhelm you with the responses. Just trying to help since these skills can be useful for contributing to other software in the future :)

**Comment by jahorta** [comment]
Okay, let me try to combine them then.

File: Source/dolphin-memory-engine.vcxproj
Comment (dreamsyntax) [inline Source/dolphin-memory-engine.vcxproj:129]: Somehow this is part of your commit.
It's already merged as part of https://github.com/aldelaro5/dolphin-memory-engine/pull/165/files

Other than that looks good now, thanks.

**Review by dreamsyntax — CHANGES_REQUESTED** [review]

File: Source/dolphin-memory-engine.vcxproj
Comment (jahorta) [reply Source/dolphin-memory-engine.vcxproj:129]: I'm sorry, I don't know how to remove this from the commit.

Comment (dreamsyntax) [reply Source/dolphin-memory-engine.vcxproj:129]: Does `git rebase master` not work?
I saw the merge commit which resolved it, but then had the merge commit.

Comment (dreamsyntax) [reply Source/dolphin-memory-engine.vcxproj:129]: If not using the dual remote strategy, go to your GitHub page:
https://github.com/jahorta/Dolphin-memory-engine/tree/master

and click "sync and DROP your own commits" (so it will revert to the current master). It should be a fairly obvious button on your page. Sync and Update? Sync and Discard? Sorry I don't recall the exact button words.

Then locally, delete your master branch:

```
git branch -D master
git fetch or git pull
git checkout master
git checkout fix_cmake_for_windows
git rebase master
git push -f
```

There are probably better/faster ways to do it, but this should resolve the issue.

Comment (jahorta) [reply Source/dolphin-memory-engine.vcxproj:129]: I tried using git rebase master when rebasing this branch and it still gave that line.

Comment (jahorta) [reply Source/dolphin-memory-engine.vcxproj]: Okay, I think I was able to get that to work. Thank you!

Comment (dreamsyntax) [reply Source/dolphin-memory-engine.vcxproj]: Looks good to me. I'll do a test locally validating the paths are good and approve if everything checks out. Thanks

File: Source/CMakeLists.txt
Comment (dreamsyntax) [reply Source/CMakeLists.txt]: Verified it is indeed correct as is.

**Review by dreamsyntax — APPROVED** [review]
Verified.

Tested:

Given Windows 11 23H2 with Visual Studio, C++ Tools & Windows SDK installed, no Qt installed on system, and the vcxproj file deleted:
- cloning this PR and NOT pulling any submodules, then running cmake locally correctly fails with:
```
CMake Error at CMakeLists.txt:56 (find_package):
  By not providing "FindQt6Widgets.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "Qt6Widgets", but CMake did not find one.

  Could not find a package configuration file provided by "Qt6Widgets" with
  any of the following names:

    Qt6WidgetsConfig.cmake
    qt6widgets-config.cmake

  Add the installation prefix of "Qt6Widgets" to CMAKE_PREFIX_PATH or set
  "Qt6Widgets_DIR" to a directory containing one of the above files.  If
  "Qt6Widgets" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
```

- pulling the Windows Qt submodule, then re-running the command successfully builds using the external as expected.

----

Thanks for doing this.
@cristian64 feel free to merge unless you have any concerns.

**Pull request merged**
